### DISCUSSION TITLE
Unable to use WASM module

### DISCUSSION THREAD

**Discussion by rygine** [MetaMask/snaps#2137] [Q&A] [answered]
source code here: https://github.com/xmtp/snap
branch: [rygine/upgrade-sdk](https://github.com/xmtp/snap/tree/rygine/upgrade-sdk)

i recently upgraded our SDK for use with a snap and the WASM integration has stopped working.

our SDK now includes 2 different ways to access the WASM module via `wasm-pack` build targets.

when using the `bundler` target (which was the previously used target), i'm seeing this error when attempting to access the WASM bindings: `MetaMask - RPC Error: Cannot read properties of undefined (reading '__wbindgen_add_to_stack_pointer')`

the snap manifest hasn't changed and includes the [webassembly endowment](https://github.com/xmtp/snap/blob/rygine/upgrade-sdk/packages/snap/snap.manifest.json#L27).

the snap config also includes the [experiment wasm support flag](https://github.com/xmtp/snap/blob/rygine/upgrade-sdk/packages/snap/snap.config.ts#L12-L14).

if i attempt to use the `web` target, which fetches the WASM module via `new URL(...)` and `fetch`, i'm seeing a different error: `MetaMask - RPC Error: Failed to construct 'URL': Invalid URL`.

for this error, the issue seems to be that `import.meta.url` is translated to a local file path: `file:///Users/user/Projects/snap/node_modules/@xmtp/user-preferences-bindings-wasm/dist/web/user_preferences_bindings_wasm.js`, which is not accessible from within the snap.

i would prefer to use the `web` targeted WASM module, but understand if that's not possible.

please advise, thanks!

**Comment by Mrtenz** [comment]
I tried to reproduce using `@xmtp/user-preferences-bindings-wasm`, but I can't get the Snap to build, probably because it's an ESM package now, whereas `xmtp-js` bundles it. Do you have a simple way to reproduce the error with `xmtp-js`?

```
✖ Compiled 1 file in 126ms with 1 error.

  • Module not found: Error: Package path ./bundler is not exported from package ./node_modules/@xmtp/user-preferences-bindings-wasm (see exports field in ./node_modules/@xmtp/user-preferences-bindings-wasm/package.json)
```

**Reply by rygine** [reply]
hmm, that error seems to be because the standard node module resolver can't resolve the `exports` field in `package.json`. have you tried changing `moduleResolution` to `Bundler`?

**Comment by rygine** [comment, marked as answer]
this has been resolved. it was an issue with `sideEffects: false` being set in `package.json`. for `wasm-pack` builds using the `bundler` option, `webpack` will tree-shake necessary code when this option is set.

also, using the `web` option in `wasm-pack` is not supported due to the use of `import.meta.url`.

thank you to @Mrtenz for help in debugging this issue!
